service/worker/scanner/tasklist/scavenger.go

```diff
@@ -154,6 +154,9 @@ func (s *Scavenger) run() {
 		s.stopWG.Done()
 	}()
 
+	// Start a task to delete orphaned tasks from the tasks table
+	s.executor.Submit(&orphanExecutorTask{scvg: s})
```

Comment:
Can you help me understand what circumstance will result in an orphan task? Thanks.

My understanding is that only this scanner will attempt to delete a task list and it only deletes a task list when:
1. The task list has been idle for a long time. This makes sure no matching hosts still owns the task list.
2. All the tasks in the tasklist have passed the expiration interval and have been successfully deleted.
3. The deletion is a conditional delete based on rangeID. If new tasks are added to the task list, rangeID will be changed and this conditional delete will fail.

Reply:
There's an important distinction here between a task being logically expired or invalid and having its database record removed. The DB record always outlasts the validity since you wouldn't want to delete a valid entry.

The TaskGC only runs as a side effect of a PollFor*Tasks call. If the poller for a task list stops polling, then TaskGC won't run and also the ACK level won't be updated. Under that circumstance (and perhaps others) eventually the task list will age out and be eligible for scavenger deletion - despite still having lingering task records which have not been handled.

It's true that all of the tasks would have passed the expiration interval by that time. But that does not imply that they have been handled. The records are still there. This is true both for scavenger when it's time to delete or for history if it updates the lease. In either case, if there are any lingering task records for the old lease, they'll still be there. History should eventually delete them via TaskGC.

The scavenger deletes the task_list entry without regard for whether (expired) task entries exist or not. So if they do, they become orphans. Previously, scavenger only iterated over the task list, so orphans were ignored and never removed.

Between this, some 300-year expiration dates holding off the scavenger, and the query bug that skips large numbers of tasks, our tasks table was over 80% orphans.

Reply:
From your above info, I think these are the two reasons that you have orphan tasks:
1.  `Previously, scavenger only iterated over the task list, so orphans were ignored and never removed`
2. ` The TaskGC only runs as a side effect of a PollFor*Tasks call. If the poller for a task list stops polling, then TaskGC won't run` 

I think 1 may be the main reason. 2 is a very edge case with much lower possibility. 

I don't understand `History should eventually delete them via TaskGC`

Also, `query bug that skips large numbers of tasks`, the bug of skipping stuff is for tasklist, not task, right? 

Correct me if I am wrong. 

What we are trying to get from here is to understand how an orphan task could occur, as our design try to guarantee this from no happening. 

However, bugs do happen sometimes, so your orphan task scavenger is important in any way.

Reply:
ahh I think I know what you meant by `History should eventually delete them via TaskGC`

What you really want to say is :
`Cadence service should eventually delete them via TaskGC`.  Current TaskGC is in Matching service, not History.